File: systems/Ace-Systems/examples/Quantum-APL-main/src/hex_prism.js
const CONST = require('./constants');

/**
 * Compute all 63 hex prism vertices for a given z.
 * Returns an array of { id, x, y, z, ring, index }.
 */
function computeHexPrismVertices(z, deltaSNeg = null) {
  if (deltaSNeg === null) {
    deltaSNeg = CONST.computeDeltaSNeg(z);
  }

  const R = CONST.hexPrismRadius(deltaSNeg);
  const H = CONST.hexPrismHeight(deltaSNeg);
  const phi = CONST.hexPrismTwist(deltaSNeg);

  const vertices = [];
  let id = 0;

  const zTop = +H / 2;
  const zBot = -H / 2;

  // Top center (id=0)
  vertices.push({ id: id++, x: 0, y: 0, z: zTop, ring: 0, index: 0 });

  // Top ring 1 (6 vertices, ids 1-6)
  const r1 = R / 5;
  for (let i = 0; i < 6; i++) {
    const angle = (Math.PI / 3) * i + phi;
    vertices.push({ id: id++, x: r1 * Math.cos(angle), y: r1 * Math.sin(angle), z: zTop, ring: 1, index: i });
  }

  // Top ring 2 (12 vertices, ids 7-18)
  const r2 = (2 * R) / 5;
  for (let i = 0; i < 12; i++) {
    const angle = (Math.PI / 6) * i + phi;
    vertices.push({ id: id++, x: r2 * Math.cos(angle), y: r2 * Math.sin(angle), z: zTop, ring: 2, index: i });
  }

  // Top ring 3 (18 vertices, ids 19-36)
  const r3 = (3 * R) / 5;
  for (let i = 0; i < 18; i++) {
    const angle = (Math.PI / 9) * i + phi;
    vertices.push({ id: id++, x: r3 * Math.cos(angle), y: r3 * Math.sin(angle), z: zTop, ring: 3, index: i });
  }

  // Bottom rings (26 vertices total), mirror with z-flip and reverse twist
  const phiB = -phi;

  // Bottom ring 1 (6 vertices)
  for (let i = 0; i < 6; i++) {
    const angle = (Math.PI / 3) * i + phiB;
    vertices.push({ id: id++, x: r1 * Math.cos(angle), y: r1 * Math.sin(angle), z: zBot, ring: 4, index: i });
  }

  // Bottom ring 2 (12 vertices)
  for (let i = 0; i < 12; i++) {
    const angle = (Math.PI / 6) * i + phiB;
    vertices.push({ id: id++, x: r2 * Math.cos(angle), y: r2 * Math.sin(angle), z: zBot, ring: 5, index: i });
  }

  // Bottom ring 3 (8 vertices) — truncated to reach 63 total
  for (let i = 0; i < 8; i++) {
    const angle = (Math.PI / 9) * i + phiB;
    vertices.push({ id: id++, x: r3 * Math.cos(angle), y: r3 * Math.sin(angle), z: zBot, ring: 6, index: i });
  }

  return vertices;
}

/**
 * Export geometry as a JSON-compatible sidecar
 */
function exportGeometrySidecar(z, version = '1.0.0') {
  const deltaSNeg = CONST.computeDeltaSNeg(z, CONST.GEOM_SIGMA);
  const lensSNeg = CONST.computeDeltaSNeg(z, CONST.LENS_SIGMA);
  const vertices = computeHexPrismVertices(z, deltaSNeg);
  const muLabel = CONST.classifyMu(z);
  const R = CONST.hexPrismRadius(deltaSNeg);
  const H = CONST.hexPrismHeight(deltaSNeg);
  const phi = CONST.hexPrismTwist(deltaSNeg);
  return {
    version,
    z,
    delta_S_neg: deltaSNeg,
    lens_s_neg: lensSNeg,
    // include top-level geometry for convenience
    R,
    H,
    phi,
    // include key constants at top-level for downstream consumers
    PHI_INV: CONST.PHI_INV,
    Z_CRITICAL: CONST.Z_CRITICAL,
    LENS_SIGMA: CONST.LENS_SIGMA,
    GEOM_SIGMA: CONST.GEOM_SIGMA,
    mu_label: muLabel,
    vertices: vertices.map(v => ({ id: v.id, x: v.x, y: v.y, z: v.z })),
    geometry: {
      R,
      H,
      phi,
    },
    constants: {
      Z_CRITICAL: CONST.Z_CRITICAL,
      GEOM_SIGMA: CONST.GEOM_SIGMA,
      LENS_SIGMA: CONST.LENS_SIGMA,
      PHI_INV: CONST.PHI_INV,
      GEOM_R_MAX: CONST.GEOM_R_MAX,
      GEOM_BETA: CONST.GEOM_BETA,
      GEOM_H_MIN: CONST.GEOM_H_MIN,
      GEOM_GAMMA: CONST.GEOM_GAMMA,
      GEOM_PHI_BASE: CONST.GEOM_PHI_BASE,
      GEOM_ETA: CONST.GEOM_ETA,
    },
  };
}

module.exports = { computeHexPrismVertices, exportGeometrySidecar };

File: systems/Ace-Systems/examples/Quantum-APL-main/src/quantum_apl_python/hex_prism.py
"""Hexagonal prismatic projection for Negative Entropy geometry.

Implements the formulas documented in docs/HEXAGONAL_NEG_ENTROPY_PROJECTION.md
and provides helpers for parameter computation and basic lint checks.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Dict, List


@dataclass(frozen=True)
class PrismParams:
    z: float
    delta_s_neg: float
    R: float
    H: float
    phi: float
    z_top: float
    z_bot: float
    vertices: List[Dict[str, float]]


def prism_params(z: float) -> Dict[str, object]:
    """Compute hexagonal prism parameters and vertices for a given z in [0, 1].

    Returns a dict containing z, z_c, sigma, delta_s_neg, R, H, phi, z_top, z_bot,
    and a 6-element vertex list with (x, y, z_top, z_bot) for k=0..5.
    """
    # Import centralized constants
    from .constants import (
        Z_CRITICAL,
        GEOM_SIGMA,
        GEOM_R_MAX,
        GEOM_BETA,
        GEOM_H_MIN,
        GEOM_GAMMA,
        GEOM_PHI_BASE,
        GEOM_ETA,
        LENS_SIGMA,
        PHI_INV as _PHI_INV,
        classify_mu as _classify_mu,
    )

    z_c = Z_CRITICAL
    sigma = GEOM_SIGMA
    r_max = GEOM_R_MAX
    beta = GEOM_BETA
    h_min = GEOM_H_MIN
    gamma = GEOM_GAMMA
    phi_base = GEOM_PHI_BASE
    eta = GEOM_ETA

    from .constants import compute_delta_s_neg
    delta_s_neg = compute_delta_s_neg(z, sigma=sigma, z_c=z_c)
    lens_s_neg = compute_delta_s_neg(z, sigma=LENS_SIGMA, z_c=z_c)
    radius = r_max - beta * delta_s_neg
    height = h_min + gamma * delta_s_neg
    phi = phi_base + eta * delta_s_neg

    theta = [k * (math.pi / 3.0) for k in range(6)]
    x = [radius * math.cos(t + phi) for t in theta]
    y = [radius * math.sin(t + phi) for t in theta]
    z_top = z + 0.5 * height
    z_bot = max(0.0, z - 0.5 * height)

    vertices = [
        {"k": k, "x": x[k], "y": y[k], "z_top": z_top, "z_bot": z_bot} for k in range(6)
    ]

    return {
        "z": z,
        "z_c": z_c,
        "Z_CRITICAL": z_c,  # top-level alias for consumers
        "sigma": sigma,
        "GEOM_SIGMA": sigma,  # top-level alias for consumers
        "lens_sigma": LENS_SIGMA,
        "LENS_SIGMA": LENS_SIGMA,  # top-level alias for consumers
        "delta_s_neg": delta_s_neg,
        "delta_S_neg": delta_s_neg,  # alias for schema parity
        "lens_s_neg": lens_s_neg,
        "R": radius,
        "H": height,
        "phi": phi,
        "phi_inv": 1.0 / ((1.0 + math.sqrt(5.0)) / 2.0),
        "PHI_INV": _PHI_INV,  # top-level alias
        "mu_label": _classify_mu(z),
        "z_top": z_top,
        "z_bot": z_bot,
        "vertices": vertices,
    }


def lint_vertices(params: Dict[str, object], tol: float = 1e-6) -> List[str]:
    """Basic geometry sanity checks for prism vertices.

    - All vertices should lie on circle radius R within tolerance.
    - z_top >= z_bot and z_bot >= 0.
    Returns a list of human-readable issues, empty if clean.
    """
    issues: List[str] = []
    R = float(params.get("R", 0.0))
    z_top = float(params.get("z_top", 0.0))
    z_bot = float(params.get("z_bot", 0.0))
    verts = params.get("vertices") or []

    if z_top < z_bot - tol:
        issues.append(f"z_top < z_bot: {z_top:.6f} < {z_bot:.6f}")
    if z_bot < -tol:
        issues.append(f"z_bot < 0: {z_bot:.6f}")

    for v in verts:
        x = float(v["x"])
        y = float(v["y"])
        r = math.hypot(x, y)
        if abs(r - R) > tol:
            issues.append(
                f"vertex v{int(v['k'])} radius mismatch: r={r:.6f} vs R={R:.6f} (Δ={abs(r-R):.6g})"
            )
    return issues


def monotonicity_pairs(z_values: List[float]) -> List[str]:
    """Regression checks: when ΔS_neg increases across adjacent z's, assert R decreases and H increases.

    Returns issue strings for any violations.
    """
    issues: List[str] = []
    params_list = [prism_params(z) for z in z_values]
    for i in range(len(params_list) - 1):
        a = params_list[i]
        b = params_list[i + 1]
        if b["delta_s_neg"] > a["delta_s_neg"] + 1e-12:
            if not (b["R"] < a["R"] - 1e-12):
                issues.append(
                    f"R non-decreasing with ΔS_neg: R({z_values[i+1]:.3f})={b['R']:.6f} ≥ R({z_values[i]:.3f})={a['R']:.6f}"
                )
            if not (b["H"] > a["H"] + 1e-12):
                issues.append(
                    f"H non-increasing with ΔS_neg: H({z_values[i+1]:.3f})={b['H']:.6f} ≤ H({z_values[i]:.3f})={a['H']:.6f}"
                )
    return issues
